### PR TITLE
lsp: Make the LSP async

### PR DESCRIPTION
**Pull request by ogoffart**

### CODE REVIEW

**Comment by ogoffart** [comment]
This is slightly more code than i thought.  
I would not be opposed to use another async lsp framework if it makes things simpler

File: tools/lsp/main.rs [hunk 1]
```diff
@@ -187,77 +197,108 @@ fn main_loop(connection: &Connection, params: serde_json::Value) -> Result<(), E
...
+                }
             }
             Message::Notification(notification) => {
-                handle_notification(&server_notifier, notification, &mut document_cache)?
+                futures.push(Box::pin(handle_notification(notification, &ctx)))
             }
         }
+
+        let mut result = Ok(());
+        futures.retain_mut(|f| {
```
Comment (tronical) [inline tools/lsp/main.rs:251]: Generally speaking, I wonder if this is the right approach. As long as the futures we generate resolve immediately on poll and all one by one then I guess it’s ok. But if any futures remain in the list, then the next time they’ll be polled is only when a new message is received. I wonder if this should either loop until all futures are resolved or use a runtime.

Comment (tronical) [reply tools/lsp/main.rs:251]: Never mind, I forgot that the whole point of the exercise is to have futures outlive the current request. Still thinking polling futures and receiver at the same time might be cleaner in the long run. But I see that we don’t need that right now :). Sorry for the false alarm.

Comment (ogoffart) [reply tools/lsp/main.rs:251]: Right now, all the future are pending on responses to requests, which happens only when we receive messages from the client and therefore get into this loop.

Comment (ogoffart) [reply tools/lsp/main.rs:251]: Perhaps this deserves a comment

**Pull request merged**